diff --git a/src/butil/iobuf.cpp b/src/butil/iobuf.cpp
index 8895fb16..a0f51fa1 100644
--- a/src/butil/iobuf.cpp
+++ b/src/butil/iobuf.cpp
@@ -1319,6 +1319,21 @@ int IOBuf::resize(size_t n, char c) {
     return 0;
 }
 
+StringPiece IOBuf::expand(size_t hint) {
+    IOBuf::Block* b = iobuf::share_tls_block();
+    if (BAIDU_UNLIKELY(!b)) {
+        StringPiece();
+    }
+
+    const size_t nc = std::min(hint, b->left_space());
+    memset(b->data + b->size, '\0', nc);
+    const IOBuf::BlockRef r = {(uint32_t)b->size, (uint32_t)nc, b};
+    _push_back_ref(r);
+    b->size += nc;
+    return StringPiece(r.block->data + r.offset, r.length);
+}
+
+
 // NOTE: We don't use C++ bitwise fields which make copying slower.
 static const int REF_INDEX_BITS = 19;
 static const int REF_OFFSET_BITS = 15;
diff --git a/src/butil/iobuf.h b/src/butil/iobuf.h
index 978f9758..c428fe74 100644
--- a/src/butil/iobuf.h
+++ b/src/butil/iobuf.h
@@ -267,6 +267,7 @@ public:
     // Returns 0 on success, -1 otherwise.
     int resize(size_t n) { return resize(n, '\0'); }
     int resize(size_t n, char c);
+    StringPiece expand(size_t hint);
 
     // Reserve `n' uninitialized bytes at back-side.
     // Returns an object representing the reserved area, INVALID_AREA on failure.
